File: dsa-problems/Arrays/solutions/ARR-012-longest-zero-sum-even.py
import sys

def longest_zero_sum_even_length(arr: list[int]) -> int:
    # Map stores {sum: {0: first_even_idx, 1: first_odd_idx}}
    # Initialize with None
    idx_map = {}

    # Base case: Sum 0 at index -1 (Odd)
    if 0 not in idx_map:
        idx_map[0] = {0: None, 1: -1}

    current_sum = 0
    max_len = 0

    for i, val in enumerate(arr):
        current_sum += val
        parity = i % 2

        if current_sum not in idx_map:
            idx_map[current_sum] = {0: None, 1: None}

        first_indices = idx_map[current_sum]

        if first_indices[parity] is not None:
            length = i - first_indices[parity]
            if length > max_len:
                max_len = length
        else:
            first_indices[parity] = i

    return max_len

def main():
    n = int(input())
    arr = list(map(int, input().split()))

    result = longest_zero_sum_even_length(arr)
    print(result)

if __name__ == "__main__":
    main()

File: dsa-problems/Arrays/solutions/java/ARR-012-longest-zero-sum-even.java
import java.util.*;

class Solution {
    public int longestZeroSumEvenLength(int[] arr) {
        // Map: Sum -> int array of size 2. [0] for even index, [1] for odd index.
        // Initialize with Integer.MIN_VALUE or similar to indicate "not seen"
        Map<Long, int[]> map = new HashMap<>();

        long currentSum = 0;
        int maxLen = 0;

        // Base case: Sum 0 at index -1. -1 is odd.
        map.put(0L, new int[]{Integer.MIN_VALUE, -1});

        for (int i = 0; i < arr.length; i++) {
            currentSum += arr[i];
            int parity = i & 1; // 0 for even, 1 for odd

            map.putIfAbsent(currentSum, new int[]{Integer.MIN_VALUE, Integer.MIN_VALUE});
            int[] firstSeen = map.get(currentSum);

            if (firstSeen[parity] != Integer.MIN_VALUE) {
                // Found a previous occurrence with same parity -> length is even
                int len = i - firstSeen[parity];
                if (len > maxLen) {
                    maxLen = len;
                }
            } else {
                // First time seeing this sum with this parity
                firstSeen[parity] = i;
            }
        }

        return maxLen;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) arr[i] = sc.nextInt();

        Solution solution = new Solution();
        int result = solution.longestZeroSumEvenLength(arr);
        System.out.println(result);
        sc.close();
    }
}
